checker/src/org/checkerframework/checker/lock/LockStore.java

```diff
@@ -136,4 +127,92 @@ protected void internalDotOutput(StringBuilder result) {
                 + "\\n");
         super.internalDotOutput(result);
     }
+
+    @Override
+    protected boolean isSideEffectFree(AnnotatedTypeFactory atypeFactory,
+            ExecutableElement method) {
+        LockAnnotatedTypeFactory lockAnnotatedTypeFactory = (LockAnnotatedTypeFactory) atypeFactory;
+        return ((LockChecker) lockAnnotatedTypeFactory.getContext()).hasOption("assumeSideEffectFree") ||
+                lockAnnotatedTypeFactory.methodSideEffectAnnotation(method, false) == SideEffectAnnotation.RELEASESNOLOCKS ||
+               super.isSideEffectFree(atypeFactory, method);
+    }
+
+    @Override
+    public void updateForMethodCall(MethodInvocationNode n,
+        AnnotatedTypeFactory atypeFactory, CFValue val) {
+        super.updateForMethodCall(n, atypeFactory, val);
+        ExecutableElement method = n.getTarget().getMethod();
+        if (!isSideEffectFree(atypeFactory, method)) {
+            // Fields are always modifiable as far as the Lock Checker is concerned, even if they are final.
+            fieldValues.clear();
+
+            // Necessary because a method could unlock a lock that is a local variable, e.g.:
+            // void foo() {
+            //     ReentrantLock lock = new ReentrantLock();
+            //     lock.lock();
+            //     unlockMyLock(lock);
+            // }
+            localVariableValues.clear();
+            // TODO: This is too conservative. Clear only the values for the local variables that could be affected.
+        }
+    }
+
+    boolean hasLockHeld(CFValue value) {
+        assert value != null;
+        AnnotatedTypeMirror type = value.getType();
+        if (type != null) {
+            AnnotationMirror anno = type.getAnnotationInHierarchy(LOCKPOSSIBLYHELD);
+            if (anno != null) {
+                return anno.equals(LOCKHELD);
```

Comment:
Isn't this whole method much easier as `return value.getType().hasAnnotation(LOCKHELD)`?